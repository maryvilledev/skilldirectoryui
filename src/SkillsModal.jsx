import React from 'react';
import axios from 'axios';
import 'react-select/dist/react-select.css';

var Select = require('react-select');

var skillTypes = [
  { value: 'compiled', label: 'Compiled' },
  { value: 'scripted', label: 'Scripted' },
  { value: 'database', label: 'Database' },
  { value: 'orchestration', label: 'Orchestration' }
];

class SkillForm extends React.Component {
  constructor(props) {
    super(props);
    this.state = {
      skill_name: "",
      skill_type: ""
    };
    this.logChange = this.logChange.bind(this);
    this.onChange = this.onChange.bind(this);
    this.onSubmit = this.onSubmit.bind(this);
  }

  logChange(val) {
    console.log("Selected:" + val);
  }

  onChange(key, value) {
    this.setState({ [key]: value});
  }

  onSubmit(ev) {
    ev.preventDefault();
    const skill_name = this.state.skill_name;
    const skill_type = this.state.skill_type;

    axios.post(this.props.api + '/skills/', {
        name: skill_name,
        skill_type: skill_type
    })
    .then(function (response) {
      console.log(response);
      this.props.closeModal();
    }.bind(this))
    .catch(err => {
      console.log('caught an error', err);
    });
  }

  render() {
    const onSkillNameChange = ev => this.onChange("skill_name", ev.target.value);
    const onSkillTypeChange = ev => this.onChange("skill_type", ev.value);
    return (
      <form onSubmit={ev => this.onSubmit(ev)}>
        <div>
          Name: <input name="skill_name" 
                       type="text" 
                       value={this.state.skill_name} 
                       onChange={onSkillNameChange} />
        </div>
        <div>
          Skill Type: <Select name="skill_type" 
                              value={this.state.skill_type} 
                              onChange={onSkillTypeChange} 
                              options={skillTypes} />
        </div>
        <button type="submit">Save</button>
      </form>
    );
  }
}

export default SkillForm
