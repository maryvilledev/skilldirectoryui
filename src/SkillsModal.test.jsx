import React from 'react';
import ReactDOM from 'react-dom';
import SkillsModal from './SkillsModal';
import { shallow } from 'enzyme';
import { mount } from 'enzyme';

describe('<SkillsModal />', () => {
  console.log("Testing logging");
  it('renders without crashing', () => {
    const div = document.createElement('div');
    ReactDOM.render(<SkillsModal />, div);
  });

  it('matches the stored snapshot', () => {
    const wrapper = shallow(<SkillsModal />);
    expect(wrapper.get(0)).toMatchSnapshot();
  });

  it('changes state when input to "Name:" field changes', () => {
    const wrapper = mount(<SkillsModal />);
    // process.stdout.write(wrapper.debug());
    // process.stdout.write(JSON.stringify(wrapper.state()));
    expect(wrapper.state().skill_name).toBe("");
    const event = {target: {value: "New Value"}};
    wrapper.find({ name: "skill_name" }).simulate("change", event);
    expect(wrapper.state().skill_name).toBe("New Value");
    // process.stdout.write(JSON.stringify(wrapper.state()));
  });

// TODO: Find different way to test this... The below test fails because
// the "Select" element we are using from 'react-select' doesn't respond
// as expected to the simulated "change" event.
  // it('changes state when selected option in "Skill Type" field changes', () => {
  //   const wrapper = mount(<SkillsModal />);
  //   expect(wrapper.state().skill_type).toBe("");
  //   const event = {target: {value: "New Value"}};
  //   wrapper.find("Select").simulate("change", event);
  //   expect(wrapper.state().skill_type).toBe("New Value");
  // });
});
