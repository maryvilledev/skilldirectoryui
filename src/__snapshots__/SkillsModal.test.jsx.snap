exports[`<SkillsModal /> matches the stored snapshot 1`] = `
ShallowWrapper {
  "complexSelector": ComplexSelector {
    "buildPredicate": [Function],
    "childrenOfNode": [Function],
    "findWhereUnwrapped": [Function],
  },
  "length": 1,
  "node": <form
    onSubmit={[Function]}>
    <div>
        Name: 
        <input
            name="skill_name"
            onChange={[Function]}
            type="text"
            value="" />
    </div>
    <div>
        Skill Type: 
        <Select
            addLabelText="Add \"{label}\"?"
            arrowRenderer={[Function]}
            autosize={true}
            backspaceRemoves={true}
            backspaceToRemoveMessage="Press backspace to remove {label}"
            clearAllText="Clear all"
            clearValueText="Clear value"
            clearable={true}
            delimiter=","
            disabled={false}
            escapeClearsValue={true}
            filterOptions={[Function]}
            ignoreAccents={true}
            ignoreCase={true}
            inputProps={Object {}}
            isLoading={false}
            joinValues={false}
            labelKey="label"
            matchPos="any"
            matchProp="any"
            menuBuffer={0}
            menuRenderer={[Function]}
            multi={false}
            name="skill_type"
            noResultsText="No results found"
            onBlurResetsInput={true}
            onChange={[Function]}
            onCloseResetsInput={true}
            openAfterFocus={false}
            optionComponent={[Function]}
            options={
                Array [
                    Object {
                      "label": "Compiled",
                      "value": "compiled",
                    },
                    Object {
                      "label": "Scripted",
                      "value": "scripted",
                    },
                    Object {
                      "label": "Database",
                      "value": "database",
                    },
                    Object {
                      "label": "Orchestration",
                      "value": "orchestration",
                    },
                  ]
            }
            pageSize={5}
            placeholder="Select..."
            required={false}
            scrollMenuIntoView={true}
            searchable={true}
            simpleValue={false}
            tabSelectsValue={true}
            value=""
            valueComponent={[Function]}
            valueKey="value" />
    </div>
    <button
        type="submit">
        Save
    </button>
</form>,
  "nodes": Array [
    <form
      onSubmit={[Function]}>
      <div>
            Name: 
            <input
                  name="skill_name"
                  onChange={[Function]}
                  type="text"
                  value="" />
      </div>
      <div>
            Skill Type: 
            <Select
                  addLabelText="Add \"{label}\"?"
                  arrowRenderer={[Function]}
                  autosize={true}
                  backspaceRemoves={true}
                  backspaceToRemoveMessage="Press backspace to remove {label}"
                  clearAllText="Clear all"
                  clearValueText="Clear value"
                  clearable={true}
                  delimiter=","
                  disabled={false}
                  escapeClearsValue={true}
                  filterOptions={[Function]}
                  ignoreAccents={true}
                  ignoreCase={true}
                  inputProps={Object {}}
                  isLoading={false}
                  joinValues={false}
                  labelKey="label"
                  matchPos="any"
                  matchProp="any"
                  menuBuffer={0}
                  menuRenderer={[Function]}
                  multi={false}
                  name="skill_type"
                  noResultsText="No results found"
                  onBlurResetsInput={true}
                  onChange={[Function]}
                  onCloseResetsInput={true}
                  openAfterFocus={false}
                  optionComponent={[Function]}
                  options={
                        Array [
                              Object {
                                "label": "Compiled",
                                "value": "compiled",
                              },
                              Object {
                                "label": "Scripted",
                                "value": "scripted",
                              },
                              Object {
                                "label": "Database",
                                "value": "database",
                              },
                              Object {
                                "label": "Orchestration",
                                "value": "orchestration",
                              },
                            ]
                  }
                  pageSize={5}
                  placeholder="Select..."
                  required={false}
                  scrollMenuIntoView={true}
                  searchable={true}
                  simpleValue={false}
                  tabSelectsValue={true}
                  value=""
                  valueComponent={[Function]}
                  valueKey="value" />
      </div>
      <button
            type="submit">
            Save
      </button>
</form>,
  ],
  "options": Object {},
  "renderer": ReactShallowRenderer {
    "_instance": ShallowComponentWrapper {
      "_calledComponentWillUnmount": false,
      "_compositeType": 0,
      "_context": Object {},
      "_currentElement": <SkillForm />,
      "_debugID": 20,
      "_hostContainerInfo": null,
      "_hostParent": null,
      "_instance": SkillForm {
        "_reactInternalInstance": [Circular],
        "context": Object {},
        "logChange": [Function],
        "onChange": [Function],
        "onSubmit": [Function],
        "props": Object {},
        "refs": Object {},
        "state": Object {
          "skill_name": "",
          "skill_type": "",
        },
        "updater": Object {
          "enqueueCallback": [Function],
          "enqueueCallbackInternal": [Function],
          "enqueueElementInternal": [Function],
          "enqueueForceUpdate": [Function],
          "enqueueReplaceState": [Function],
          "enqueueSetState": [Function],
          "isMounted": [Function],
          "validateCallback": [Function],
        },
      },
      "_mountOrder": 5,
      "_pendingCallbacks": null,
      "_pendingElement": null,
      "_pendingForceUpdate": false,
      "_pendingReplaceState": false,
      "_pendingStateQueue": null,
      "_renderedComponent": NoopInternalComponent {
        "_currentElement": <form
          onSubmit={[Function]}>
          <div>
                    Name: 
                    <input
                              name="skill_name"
                              onChange={[Function]}
                              type="text"
                              value="" />
          </div>
          <div>
                    Skill Type: 
                    <Select
                              addLabelText="Add \"{label}\"?"
                              arrowRenderer={[Function]}
                              autosize={true}
                              backspaceRemoves={true}
                              backspaceToRemoveMessage="Press backspace to remove {label}"
                              clearAllText="Clear all"
                              clearValueText="Clear value"
                              clearable={true}
                              delimiter=","
                              disabled={false}
                              escapeClearsValue={true}
                              filterOptions={[Function]}
                              ignoreAccents={true}
                              ignoreCase={true}
                              inputProps={Object {}}
                              isLoading={false}
                              joinValues={false}
                              labelKey="label"
                              matchPos="any"
                              matchProp="any"
                              menuBuffer={0}
                              menuRenderer={[Function]}
                              multi={false}
                              name="skill_type"
                              noResultsText="No results found"
                              onBlurResetsInput={true}
                              onChange={[Function]}
                              onCloseResetsInput={true}
                              openAfterFocus={false}
                              optionComponent={[Function]}
                              options={
                                        Array [
                                                  Object {
                                                    "label": "Compiled",
                                                    "value": "compiled",
                                                  },
                                                  Object {
                                                    "label": "Scripted",
                                                    "value": "scripted",
                                                  },
                                                  Object {
                                                    "label": "Database",
                                                    "value": "database",
                                                  },
                                                  Object {
                                                    "label": "Orchestration",
                                                    "value": "orchestration",
                                                  },
                                                ]
                              }
                              pageSize={5}
                              placeholder="Select..."
                              required={false}
                              scrollMenuIntoView={true}
                              searchable={true}
                              simpleValue={false}
                              tabSelectsValue={true}
                              value=""
                              valueComponent={[Function]}
                              valueKey="value" />
          </div>
          <button
                    type="submit">
                    Save
          </button>
</form>,
        "_debugID": 21,
        "_renderedOutput": <form
          onSubmit={[Function]}>
          <div>
                    Name: 
                    <input
                              name="skill_name"
                              onChange={[Function]}
                              type="text"
                              value="" />
          </div>
          <div>
                    Skill Type: 
                    <Select
                              addLabelText="Add \"{label}\"?"
                              arrowRenderer={[Function]}
                              autosize={true}
                              backspaceRemoves={true}
                              backspaceToRemoveMessage="Press backspace to remove {label}"
                              clearAllText="Clear all"
                              clearValueText="Clear value"
                              clearable={true}
                              delimiter=","
                              disabled={false}
                              escapeClearsValue={true}
                              filterOptions={[Function]}
                              ignoreAccents={true}
                              ignoreCase={true}
                              inputProps={Object {}}
                              isLoading={false}
                              joinValues={false}
                              labelKey="label"
                              matchPos="any"
                              matchProp="any"
                              menuBuffer={0}
                              menuRenderer={[Function]}
                              multi={false}
                              name="skill_type"
                              noResultsText="No results found"
                              onBlurResetsInput={true}
                              onChange={[Function]}
                              onCloseResetsInput={true}
                              openAfterFocus={false}
                              optionComponent={[Function]}
                              options={
                                        Array [
                                                  Object {
                                                    "label": "Compiled",
                                                    "value": "compiled",
                                                  },
                                                  Object {
                                                    "label": "Scripted",
                                                    "value": "scripted",
                                                  },
                                                  Object {
                                                    "label": "Database",
                                                    "value": "database",
                                                  },
                                                  Object {
                                                    "label": "Orchestration",
                                                    "value": "orchestration",
                                                  },
                                                ]
                              }
                              pageSize={5}
                              placeholder="Select..."
                              required={false}
                              scrollMenuIntoView={true}
                              searchable={true}
                              simpleValue={false}
                              tabSelectsValue={true}
                              value=""
                              valueComponent={[Function]}
                              valueKey="value" />
          </div>
          <button
                    type="submit">
                    Save
          </button>
</form>,
      },
      "_renderedNodeType": 0,
      "_rootNodeID": 0,
      "_topLevelWrapper": null,
      "_updateBatchNumber": null,
      "_warnedAboutRefsInRender": false,
    },
    "getRenderOutput": [Function],
    "render": [Function],
  },
  "root": [Circular],
  "unrendered": <SkillForm />,
}
`;
